library/kani_macros/src/sysroot/contracts/shared.rs

```diff
@@ -176,3 +179,246 @@ pub fn try_as_result_assign(stmt: &syn::Stmt) -> Option<&syn::LocalInit> {
+    let mut remembers_exprs = HashMap::new();
+    let mut vis =
+        OldVisitor { t: OldLifter::new(), remember_count, remembers_exprs: &mut remembers_exprs };
+    let mut expr = &mut data.clone();
+    vis.visit_expr_closure_mut(&mut expr);
+
+    let arg_names = rename_argument_occurrences(fn_sig, &mut expr);
+    let (start, end) = make_unsafe_argument_copies(&arg_names);
+
+    let remembers_stmts: TokenStream2 = remembers_exprs
+        .iter()
+        .fold(start, |collect, (ident, expr)| quote!(let #ident = #expr; #collect));
+
+    let result: Ident = Ident::new(INTERNAL_RESULT_IDENT, Span::call_site());
+    (remembers_stmts, end, Expr::Verbatim(quote!((#expr)(&#result))))
+}
+
+trait OldTrigger {
+    /// You are provided the expression that is the first argument of the
+    /// `old()` call. You may modify it as you see fit. The return value
+    /// indicates whether the entire `old()` call should be replaced by the
+    /// (potentially altered) first argument.
+    ///
+    /// The second argument is the span of the original `old` expr.
+    ///
+    /// The third argument is the number of remember variables that have already been
+    /// instantiated in the surrounding environment.
+    ///
+    /// The fourth argument is a collection of all the expressions that need to be lifted
+    /// into the past environment as new remember variables.
+    fn trigger(
+        &mut self,
+        e: &mut Expr,
+        s: Span,
+        remember_count: usize,
+        output: &mut HashMap<Ident, Expr>,
+    ) -> bool;
+}
+
+struct OldLifter;
+
+impl OldLifter {
+    fn new() -> Self {
+        Self
+    }
+}
+
+struct OldDenier;
+
+impl OldTrigger for OldDenier {
+    fn trigger(&mut self, _: &mut Expr, s: Span, _: usize, _: &mut HashMap<Ident, Expr>) -> bool {
+        s.unwrap().error("Nested calls to `old` are prohibited").emit();
+        false
+    }
+}
+
+struct OldVisitor<'a, T> {
+    t: T,
+    remember_count: usize,
+    remembers_exprs: &'a mut HashMap<Ident, Expr>,
+}
+
+impl<T: OldTrigger> syn::visit_mut::VisitMut for OldVisitor<'_, T> {
+    fn visit_expr_mut(&mut self, ex: &mut Expr) {
+        let trigger = match &*ex {
+            Expr::Call(call @ ExprCall { func, attrs, args, .. }) => match func.as_ref() {
+                Expr::Path(ExprPath {
+                    attrs: func_attrs,
+                    qself: None,
+                    path: Path { leading_colon: None, segments },
+                }) if segments.len() == 1
+                    && segments.first().map_or(false, |sgm| sgm.ident == "old") =>
+                {
+                    let first_segment = segments.first().unwrap();
+                    assert_spanned_err!(first_segment.arguments.is_empty(), first_segment);
+                    assert_spanned_err!(attrs.is_empty(), call);
+                    assert_spanned_err!(func_attrs.is_empty(), func);
+                    assert_spanned_err!(args.len() == 1, call);
```

Comment:
Let's add a string description for each assertion to provide a better error message.

Reply:
Issue mentioned in #3273